wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,299 @@
+    T: Fn() -> Result<S, S::Error> + Send + Sync + 'static
+{
+}
+
+/// Count erroneous attempts while quering some data source and perform reinitialization/fallback.
+///
+/// To use within an object, you must implement `FallibleDataSource` first.
+///
+/// Example:
+/// ```rust
+/// fn main() {
+///     struct Repo;
+///     struct RepoError;
+///
+///     impl FallibleDataSource for Repo {
+///         type Error = RepoError;
+///
+///         fn is_countable_err(err: &Self::Error) -> bool {
+///             true
+///         }
+///     }
+///
+///     let cb = CircuitBreaker::builder()
+///         .with_max_timespan(Duration::from_secs(1))
+///         .with_max_err_count_per_timespan(5)
+///         .with_init_fn(|| Ok(Repo));
+///
+///     cb.query(|src| async move { Err(RepoError) }).unwrap_err()
+///     cb.query(|src| async move { Ok(()) }).unwrap()
+///     
+///     // see CB test for more verbose example
+/// }
+/// ```
+pub struct CircuitBreaker<S: FallibleDataSource> {
+    /// Timespan that errors will be counted in.
+    /// After it elapsed, error counter will be resetted.
+    max_timespan: Duration,
+
+    /// Maximum error count per timespan. Example: 3 errors per 1 sec (max_timespan)
+    max_err_count_per_timespan: usize,
+
+    /// A function that may be called on every fail to reinitialize data source
+    init_fn: Box<dyn DataSrcInitFn<S>>,
+
+    /// Current state of CB
+    state: RwLock<CBState<S>>,
+}
+
+struct CBState<S: FallibleDataSource> {
+    data_source: Arc<S>,
+    err_count: usize,
+    first_err_ts: Option<Instant>,
+}
+
+impl<S: FallibleDataSource> CBState<S> {
+    fn inc(&mut self) {
+        self.err_count += 1;
+    }
+
+    fn reset(&mut self) {
+        self.err_count = 0;
+        self.first_err_ts = None;
+    }
+
+    fn reinit(&mut self, src: S) {
+        self.data_source = Arc::new(src);
+    }
+}
+
+pub struct CircuitBreakerBuilder<S: FallibleDataSource> {
+    max_timespan: Option<Duration>,
+    max_err_count_per_timespan: Option<usize>,
+    init_fn: Option<Box<dyn DataSrcInitFn<S>>>,
+}
+
+impl<S: FallibleDataSource> CircuitBreakerBuilder<S> {
+    pub fn new() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder {
+            max_timespan: None,
+            max_err_count_per_timespan: None,
+            init_fn: None,
+        }
+    }
+
+    pub fn with_max_timespan(mut self, ts: Duration) -> CircuitBreakerBuilder<S> {
+        self.max_timespan = Some(ts);
+        self
+    }
+
+    pub fn with_max_err_count_per_timespan(mut self, count: usize) -> CircuitBreakerBuilder<S> {
+        self.max_err_count_per_timespan = Some(count);
+        self
+    }
+
+    pub fn with_init_fn(mut self, f: impl DataSrcInitFn<S>) -> CircuitBreakerBuilder<S> {
+        self.init_fn = Some(Box::new(f));
+        self
+    }
+
+    /// Build the circuit breaker.
+    /// Note: you should set up all 3 fields within this builder, either it will panic
```

Comment:
If so, better not to use Builder patter at all, it doesn't fit well here. A simple struct with all fields public would do.